### PR TITLE
Incremental find

### PR DESCRIPTION
**Pull request by scholtzan**
# Incremental Find

## Summary

This implements incremental search. When doing find on a large document then find is performed on batches of document regions. The process starts with the visible document range and expands from there.

## Motivation

With the current implementation of find, searching large files can be very slow. One reason for this is that for each find request all occurrences are deterimined by core and then sent to the frontend. Users might have to wait for a long time to see any of their search results.

This PR tries to solve this problem by searching text incrementally. This means that find queries are executed on smaller ranges of the text and matches are reported to the frontend after each range has been searched. 

## How it currently works

Right now, incremental search starts with executing find on the visible range, since this would mean that users see visible results quickly. After searching the visible range, find will continue searching the ranges that come before and after the visible range and continues doing that until the entire text has been searched. These ranges will overlap with eachother so that also matches will be found that might otherwise be split over two ranges.

This approach works for simple search queries and regex queries, however it does not work for regex that could match multiple lines (multi-line regex). To also support multi-line regexes, these will be executed on the entire text after the previously described incremental find has searched the entire text.  

## Implementation

The implementation is similar to word wrapping. Find operations are scheduled and handled during idle. After each `do_incremental_find`, `find_status` which contains all matches found up until this point is sent to the frontend.

`find_progress` has been added to `View` to keep track of whether an incremental find is currently running:

```rust
enum FindProgress {
    /// Incremental find is done/not running.
    Done,

    /// Incremental find is in progress. Keeps tracked of already searched range.
    InProgress(Option<Range<usize>>),
}
```

Most of the relevant find logic is either in `view.rs` `do_find` and `set_find` and in `find.rs`. 

I am also considering to move the find logic currently in `View` into a separate object (`Find` then the current `Find` would become `FindQuery`) to make the code a bit cleaner.

I will also add tests to ensure that everything is working as it's supposed to. Right now there are a few bugs which I'll fix, but first I wanted to make sure that the current approach goes into the right direction.

## Todo

- [x] Incremental search
- [x] Cleanup logic
- [x] Make it work with multi line regexes
- [x] Fix bugs
    - [x] Do find on pasted text (fixed in https://github.com/xi-editor/xi-editor/pull/1103)
    - [x] `find_status` after edit for update find results
- [x] Tests

## Review Checklist

- [x] I have responded to reviews and made changes where appropriate.
- [x] I have tested the code with `cargo test --all` / `./rust/run_all_checks`.
- [x] I have updated comments / documentation related to the changes I made.
- [x] I have rebased my PR branch onto xi-editor/master.

### CODE REVIEW

File: rust/core-lib/src/view.rs [hunk 1]
```diff
@@ -107,6 +113,19 @@ enum FindStatusChange {
...
+#[derive(PartialEq, Debug, Clone)]
+enum FindProgress {
+    /// Incremental find is done/not running.
+    Done,
+
+    /// Incremental find is in progress. Keeps tracked of already searched range.
+    InProgress(Range<usize>),
+
+    /// Incremental find process just started.
+    Started,
```
[the remark this answers is not in the document]
Comment (scholtzan) [reply rust/core-lib/src/view.rs]: Oh well, that would also work

**Comment by scholtzan** [comment]
This is ready for review now. I added a bunch of tests for find in general and incremental find.

**Comment by scholtzan** [comment]
It would probably also make sense to open an issue regarding the relationship between the incremental approach and trying to do tasks like this on a background thread (basically what was mentioned on zulip https://xi.zulipchat.com/#narrow/stream/147924-xi-editor/topic/xi.231083/near/156396552)?

File: rust/core-lib/src/view.rs [hunk 2]
```diff
@@ -107,6 +113,17 @@ enum FindStatusChange {
...
 
+/// Indicates what changed in the find state.
+#[derive(PartialEq, Debug, Clone)]
+enum FindProgress {
+    /// Incremental find is done/not running.
+    Done,
+
+    /// Incremental find is in progress. Keeps tracked of already searched range.
+    /// `None` if find process just started.
+    InProgress(Option<Range<usize>>),
```
[the remark this answers is not in the document]
Comment (scholtzan) [reply rust/core-lib/src/view.rs]: Yeah, you suggested to use an empty range here. But I think I prefer having `Started` back as a state

File: rust/core-lib/src/find.rs [hunk 3]
```diff
@@ -440,18 +430,165 @@ mod tests {
     fn find() {
         let base_text = Rope::from("hello world");
         let mut find = Find::new(1);
-        find.do_find(&base_text, "world", false, false, false);
+        find.set_find("world", false, false, false);
+        find.update_find(&base_text, 0, base_text.len(), false);
         assert_eq!(find.occurrences().len(), 1);
         assert_eq!(find.occurrences().first(), Some(&SelRegion::new(6, 11)));
+    }
+
```
[the remark this answers is not in the document]
Comment (scholtzan) [reply rust/core-lib/src/find.rs]: they are, no?

File: rust/core-lib/src/event_context.rs [hunk 4]
```diff
@@ -140,6 +142,9 @@ impl<'a> EventContext<'a> {
                 if self.with_view(|v, t| v.needs_wrap_in_visible_region(t)) {
                     self.rewrap();
                 }
+                if self.with_view(|v, _| v.find_in_progress()) {
```
[the remark this answers is not in the document]
Comment (scholtzan) [reply rust/core-lib/src/event_context.rs]: Not sure if there is something similar for find. Right now it is the case that it only executes find if the query actually changed. In that case `find_in_progress` would be `true`. So in a way it only does find if necessary.

**Comment by scholtzan** [comment]
There was already a `TODO` in the find code to prevent cutting codepoints. Should be fixed now. I also added a test for it.

File: rust/core-lib/src/find.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmyr) [reply rust/core-lib/src/find.rs]: Oops yep my bad 😅

**Comment by cmyr** [comment]
A question I'm not able to resolve by scanning through this: What happens if an edit occurs while find is in progress?

**Comment by scholtzan** [comment]
If an edit happens then find is applied on the delta (for deletions occurrences are deleted, for insertion find is applied on the inserted text + some slope). So, I believe it doesn't interfere with incremental find. If the edit happens in the already searched range while incremental find is still going on, then the find results occurring there will just get updated. If it happens outside the already searched range (although should be very unlikely to happen) then the find results there will get overwritten by incremental find later by what should be the same results.

File: rust/core-lib/src/view.rs [hunk 5]
```diff
@@ -1021,14 +1048,80 @@ impl View {
...
+                    // stop incremental find
+                    self.find_progress = FindProgress::Ready;
+                    Some((0, text.len()))
+                } else {
+                    // expand find to un-searched regions
+                    let start_off = self.offset_of_line(text, self.first_line);
+
+                    // check if preceding range should be searched next
+                    let search_preceding_range =
+                        start_off.checked_sub(searched_range.start).unwrap_or_else(|| 0)
```
[the remark this answers is not in the document]
Comment (scholtzan) [reply rust/core-lib/src/view.rs]: Cool, didn't know that existed!

File: rust/core-lib/src/view.rs [hunk 6]
```diff
@@ -1021,14 +1048,80 @@ impl View {
...
+                    // the entire text has been searched
+                    // end find by executing multi-line regex queries on entire text
+                    // stop incremental find
+                    self.find_progress = FindProgress::Ready;
+                    Some((0, text.len()))
+                } else {
+                    // expand find to un-searched regions
+                    let start_off = self.offset_of_line(text, self.first_line);
+
+                    // check if preceding range should be searched next
```
[the remark this answers is not in the document]
Comment (scholtzan) [reply rust/core-lib/src/view.rs]: Yeah, that sounds easier to understand.

File: rust/core-lib/src/find.rs [hunk 7]
```diff
@@ -270,13 +260,23 @@ impl Find {
         let search_string = self.search_string.as_ref().unwrap();
 
         // expand region to be able to find occurrences around the region's edges
-        let from = max(start, slop) - slop;
-        let to = min(end + slop, text.len());
+        let expanded_start = max(start, slop) - slop;
+        let expanded_end = min(end + slop, text.len());
+
+        let from = if text.is_codepoint_boundary(expanded_start) {
```
[the remark this answers is not in the document]
Comment (scholtzan) [reply rust/core-lib/src/find.rs]: I have seen somethings like this in `resolve_offset` in `plugin/rpc.rs`, so probably a good idea to have a function for this.

File: rust/core-lib/src/view.rs [hunk 8]
```diff
@@ -1021,14 +1048,80 @@ impl View {
...
                 query.regex,
                 query.whole_words,
             )
         }
+
+        self.find_progress = FindProgress::Started;
+    }
+
+    pub fn do_find(&mut self, text: &Rope) {
+        let search_range = match &self.find_progress.clone() {
```
[the remark this answers is not in the document]
Comment (scholtzan) [reply rust/core-lib/src/view.rs]: `Range` doesn't implement `Copy` (https://stackoverflow.com/questions/43416914/why-doesnt-opsranget-implement-copy-even-if-t-is-copy)

**Comment by scholtzan** [comment]
I added some tracing and most of the time is spend on sending `find_status` to the frontend after each stage of incremental find. Probably because it contains a list of all matches. I guess those can be removed since they are also sent when updating annotations.
I also did some profiling to determine a good `FIND_BATCH_SIZE`. 

One thing I noticed is that line wrapping is quite slow while incremental search is in progress.

**Comment by scholtzan** [comment]
Odd... It seems like when doing `text.subseq` at some point `push_subseq` cuts through a codepoint and not at a codepoint boundary.

**Comment by cmyr** [comment]
Yea, I also noticed linewrap seeming slow in this situation. I'm curious about what's going on, there. It might be that there's contention for `idle`?

This makes sense that we don't want to be sending a big list of matches unnecessarily or redundantly. I'm open to ideas that can reduce this overhead? This does get tricky.

**Comment by scholtzan** [comment]
I found the cause for the slow wrapping. `render` sent `find_status` every time it got called although it would have been sufficient to send the changes just once since they don't change. I fixed that.

**Comment by scholtzan** [comment]
Okay, I made some performance improvements. Now all the line occurrences in `find_status` get send only once find is done. Before it was send after every iteration which slowed it down a bit.

**Comment by scholtzan** [comment]
I added some documentation.

There was also a frontend bug which made the case sensitivity change in the background: https://github.com/xi-editor/xi-mac/pull/445

**Pull request merged**